services/worker-manager/test/provider_azure_test.js

```diff
@@ -0,0 +1,301 @@
+    await worker.reload();
+    assert.equal(worker.state, helper.Worker.states.REQUESTED); // RUNNING is set by register which does not happen here
+
+    // And now we fake it is stopped
+    await provider.scanPrepare();
+    await provider.checkWorker({worker});
+    await provider.scanCleanup();
+    await worker.reload();
+    assert.equal(worker.state, helper.Worker.states.STOPPED);
+  });
+
+  test('update long-running worker', async function() {
+    const expires = taskcluster.fromNow('-1 week');
+    const worker = await helper.Worker.create({
+      workerPoolId,
+      workerGroup: 'whatever',
+      workerId: 'whatever',
+      providerId,
+      created: taskcluster.fromNow('-2 weeks'),
+      lastModified: taskcluster.fromNow('-2 weeks'),
+      lastChecked: taskcluster.fromNow('-2 weeks'),
+      capacity: 1,
+      expires,
+      state: helper.Worker.states.RUNNING,
+      providerData: {
+        ...baseProviderData,
+      },
+    });
+    await provider.scanPrepare();
+    await provider.checkWorker({worker});
+    await provider.scanCleanup();
+    assert(worker.expires > expires);
+  });
+
+  test('remove unregistered workers', async function() {
+    const worker = await helper.Worker.create({
+      workerPoolId,
+      workerGroup: 'whatever',
+      workerId: 'whatever',
+      providerId,
+      capacity: 1,
+      created: taskcluster.fromNow('-1 hour'),
+      lastModified: taskcluster.fromNow('-2 weeks'),
+      lastChecked: taskcluster.fromNow('-2 weeks'),
+      expires: taskcluster.fromNow('1 week'),
+      state: helper.Worker.states.REQUESTED,
+      providerData: {
+        ...baseProviderData,
+        registrationExpiry: Date.now() - 1000,
+      },
+    });
+    await provider.scanPrepare();
+    await provider.checkWorker({worker});
+    await provider.scanCleanup();
+    assert(fakeAzure.deleteVMStub.called);
+    assert(fakeAzure.deleteDiskStub.called);
+    assert(fakeAzure.deleteIPStub.called);
+    assert(fakeAzure.deleteNICStub.called);
+  });
+
+  test('don\'t remove unregistered workers that are new', async function() {
+    const worker = await helper.Worker.create({
+      workerPoolId,
+      workerGroup: 'whatever',
+      workerId: 'whatever',
+      providerId,
+      created: taskcluster.fromNow('-1 hour'),
+      expires: taskcluster.fromNow('1 week'),
+      capacity: 1,
+      lastModified: taskcluster.fromNow('-2 weeks'),
+      lastChecked: taskcluster.fromNow('-2 weeks'),
+      state: helper.Worker.states.REQUESTED,
+      providerData: {
+        ...baseProviderData,
+        registrationExpiry: Date.now() + 1000,
+      },
+    });
+    await provider.scanPrepare();
+    await provider.checkWorker({worker});
+    await provider.scanCleanup();
+    assert(!fakeAzure.deleteVMStub.called);
+    assert(!fakeAzure.deleteDiskStub.called);
+    assert(!fakeAzure.deleteIPStub.called);
+    assert(!fakeAzure.deleteNICStub.called);
+  });
+
+  suite('registerWorker', function() {
```

Comment:
Similar to the AWS tests (https://github.com/taskcluster/taskcluster/blob/master/services/worker-manager/test/provider_aws_test.js#L257) it would be nice to see some failure tests with bad azure signatures, incorrect vmId, and maybe bad encoding and bad chain of trust.

If possible, should aim for near 100% test coverage of this function.